File: src/imagebuilder/flyingtux/app/builder.py
from metux.util.specobject import SpecObject
from metux.util.log import info
from ..pkg import get as pkg_get
from ..buildjail import get as buildjail_get
from toolbase import ToolBase

# 2do: * fetch arch from target
#      * check what else can be fetched from target
class Builder(ToolBase):
    def __init__(self, spec):
        ToolBase.__init__(self, spec, 'Builder')

    """[private]"""
    def __do_single_os_component(self, name, component):
        pkg = component['packages']
        if pkg is not None:
            self.info("osbase component %s: installing packages: %s" % (name,str(pkg)))
            self.my_pkg.add_packages(pkg)

        selector = component['selector']
        if selector is not None:
            choice = component['choice::'+selector]
            if choice is None:
                self.info("osbase component %s: cant find choice %s - using default" % (name, selector))
                choice = component['default']

            if choice is not None:
                pkg = choice['packages']
                if pkg is not None:
                    self.info("osbase component %s choice: installing packages: %s" % (name,str(pkg)))
                    self.my_pkg.add_packages(pkg)
            else:
                self.info("osbase component %s has no default choice" % name)

    """[private]"""
    def __do_os_components(self):
        for name in self.get_cf_list('IMAGE::rootfs::os-components'):
            component = self['IMAGE::OSBASE'].get_component(name)
            self.info("processing component: "+name)
            if component is None:
                self.abort("missing os component %s needed by image" % name)
            self.__do_single_os_component(name, component)

    def init_jail(self):
        # fixme: can't use default_set(), since lambdadict defaults aren't
        # automatically imported when creating SpecObject from a lambdadict
        jailconf = self['IMAGE::OSBASE::build-jail']
        jailconf['target-docker-image'] = self['DOCKER-IMAGE']
        jailconf['target-tarball']      = self['ROOTFS-TARBALL']
        jailconf['workdir']             = self['BUILD-SYSROOT']

        self.my_jail = buildjail_get(jailconf)

    def init_pkg(self):
        self.my_pkg = pkg_get(
            self['IMAGE::OSBASE::engine'],
            {
                'arch':    self['ARCH'],
            },
            self.my_jail
        )

    def build_image(self):

        self.init_jail()
        self.init_pkg()

        self.info("preparing image")
        self.my_pkg.prepare()

        self.info("adding repos")
        self.my_pkg.add_repos(self['IMAGE'].get_repos())

        self.info("adding os components")
        self.__do_os_components()

        self.info("installing application packages")
        self.my_pkg.add_packages(self['IMAGE::rootfs::packages'])

        self.info("cleaning up")
        self.my_jail.img_remove_recursive(self['IMAGE'].get_purged_files())
        self.my_jail.img_remove_dirs(self['IMAGE'].get_empty_dirs())

        self.info("finishing image")
        self.my_pkg.finish()

        self.info("image build done")
        return 0

    def run(self):
        return self.build_image()

File: internal/app/builder.go
package app

import (
	"github.com/metux/flyingtux/internal/buildjail"
	"github.com/metux/flyingtux/internal/fs"
	"github.com/metux/flyingtux/internal/pkg"
	"github.com/metux/flyingtux/internal/spec"
)

// Builder assembles an image (port of app.builder.Builder).
type Builder struct {
	toolBase
	image    *spec.SpecObj
	platform *spec.SpecObj
	jail     buildjail.BuildJail
	pkg      pkg.Pkg
}

// Run performs the full image build (port of Builder.run).
func (b *Builder) Run() (int, error) {
	platform := b.Sub("PLATFORM")
	image := b.Sub("IMAGE")
	osbase := image.Sub("OSBASE")

	// object fixups: make ${PLATFORM::..} resolvable within image/osbase
	image.SetEntry("PLATFORM", platform.Entry)
	osbase.SetEntry("PLATFORM", platform.Entry)

	b.info("initializing build jail")
	buildTemp := b.GetStr("BUILD-TEMP")
	_ = fs.Rmtree(buildTemp)

	jailEngine := b.GetStr("TARGET::jail-engine")
	jail, err := buildjail.Get(buildjail.Params{
		RootfsImage: b.GetStr("ROOTFS-IMAGE"),
		Workdir:     buildTemp,
		Engine:      jailEngine,
		ConfImage:   osbase.GetStr(spec.Key("build-jail::" + jailEngine + "::image")),
		ConfInit:    osbase.GetStr(spec.Key("build-jail::" + jailEngine + "::init")),
	})
	if err != nil {
		return 0, err
	}
	b.jail = jail

	p, err := pkg.Get(osbase.GetStr("engine"), b.GetStr("ARCH"), jail)
	if err != nil {
		return 0, err
	}
	b.pkg = p

	b.info("preparing image")
	if err := p.Prepare(); err != nil {
		return 0, err
	}

	b.info("adding repos")
	r := imageRepos(image)
	if err := p.AddRepos(&pkg.Repos{URLs: r.URLs, Keys: r.Keys}); err != nil {
		return 0, err
	}

	b.info("copying base files")
	for _, e := range osbase.SubElems("rootfs::copy-files") {
		ent := spec.Wrap(e)
		if err := jail.SysrootCopyFile(ent.GetStr("source"), ent.GetStr("dest"), ent.GetStr("mode"), ent.GetStr("owner")); err != nil {
			return 0, err
		}
	}

	if err := jail.SysrootCreateDirs(append(
		osbase.GetStrList("rootfs::create-dirs"),
		image.GetStrList("rootfs::create-dirs")...)); err != nil {
		return 0, err
	}

	for _, e := range image.SubElems("rootfs::create-links") {
		ent := spec.Wrap(e)
		if err := jail.SysrootSymlink(ent.GetStr("name"), ent.GetStr("target")); err != nil {
			return 0, err
		}
	}

	b.info("adding os components")
	if err := b.doOSComponents(image.GetStrList("rootfs::os-components"), osbase); err != nil {
		return 0, err
	}

	b.info("installing application packages")
	if err := p.AddPackages(image.GetStrList("rootfs::packages")); err != nil {
		return 0, err
	}

	b.info("cleaning up")
	if err := jail.SysrootRemoveRecursive(purgedFiles(image)); err != nil {
		return 0, err
	}
	if err := jail.SysrootRemoveDirs(emptyDirs(image)); err != nil {
		return 0, err
	}

	b.info("finishing image")
	if err := p.Finish(); err != nil {
		return 0, err
	}

	b.info("cleaning working directory: " + buildTemp)
	_ = fs.Rmtree(buildTemp)

	b.info("image build done")
	return 0, nil
}

// doOSComponents processes a list of osbase components recursively (port of
// __do_os_components).
func (b *Builder) doOSComponents(names []string, osbase *spec.SpecObj) error {
	for _, name := range names {
		key := spec.Key("components::" + name)
		if !osbase.Has(key) {
			return b.abort("missing os component " + name + " needed by image")
		}
		b.info("processing component: " + name)
		if err := b.doSingleComponent(name, osbase.Sub(key), osbase); err != nil {
			return err
		}
	}
	return nil
}

// doSingleComponent installs a component's packages and selected choice (port
// of __do_single_os_component).
func (b *Builder) doSingleComponent(name string, comp, osbase *spec.SpecObj) error {
	if err := b.doOSComponents(comp.GetStrList("depends"), osbase); err != nil {
		return err
	}

	if pkgs := comp.GetStrList("packages"); len(pkgs) > 0 {
		b.info("osbase component " + name + ": installing packages")
		if err := b.pkg.AddPackages(pkgs); err != nil {
			return err
		}
	}

	selector := comp.GetStr("selector")
	if selector == "" {
		return nil
	}

	choiceKey := spec.Key("choice::" + selector)
	var choice *spec.SpecObj
	if comp.Has(choiceKey) {
		choice = comp.Sub(choiceKey)
	} else if comp.Has("default") {
		b.info("osbase component " + name + ": cant find choice " + selector + " - using default")
		choice = comp.Sub("default")
	} else {
		b.info("osbase component " + name + " has no default choice")
		return nil
	}

	if pkgs := choice.GetStrList("packages"); len(pkgs) > 0 {
		b.info("osbase component " + name + " choice: installing packages")
		if err := b.pkg.AddPackages(pkgs); err != nil {
			return err
		}
	}
	return nil
}

func (b *Builder) abort(text string) error {
	return &abortErr{b.toolname + ": " + text}
}

type abortErr struct{ msg string }

func (e *abortErr) Error() string { return e.msg }
